spec.html

```diff
@@ -21293,10 +21289,15 @@ <h1>Runtime Semantics: Evaluation</h1>
       <emu-note>
         <p><emu-grammar>ClassDeclaration : `class` ClassTail</emu-grammar> only occurs as part of an |ExportDeclaration| and is never directly evaluated.</p>
       </emu-note>
-      <emu-grammar>ClassExpression : `class` BindingIdentifier? ClassTail</emu-grammar>
+      <emu-grammar>ClassExpression : `class` ClassTail</emu-grammar>
       <emu-alg>
-        1. If |BindingIdentifier_opt| is not present, let _className_ be *undefined*.
-        1. Else, let _className_ be StringValue of |BindingIdentifier|.
+        1. Let _value_ be ? ClassDefinitionEvaluation of |ClassTail| with arguments *undefined* and *""*.
+        1. Set _value_.[[SourceText]] to the source text matched by |ClassExpression|.
+        1. Return _value_.
+      </emu-alg>
+      <emu-grammar>ClassExpression : `class` BindingIdentifier ClassTail</emu-grammar>
```

Comment:
Hm, I think we don't typically split syntax-directed operations on presence/absence of an optional nonterminal like this. Maybe I'm wrong, but if not, this should be just one SDO along the lines of

```html
      <emu-grammar>ClassExpression : `class` BindingIdentifier? ClassTail</emu-grammar>
        1. If |BindingIdentifier| is present, then
          1. Let _className_ be StringValue of |BindingIdentifier|.
          1. Let _value_ be ? ClassDefinitionEvaluation of |ClassTail| with arguments _className_ and _className_.
        1. Else,
          1. Let _value_ be ? ClassDefinitionEvaluation of |ClassTail| with arguments *undefined* and *""*.
        1. Set _value_.[[SourceText]] to the source text matched by |ClassExpression|.
        1. Return _value_.
      </emu-grammar>
```

Reply:
> I think we don't typically split syntax-directed operations on presence/absence of an optional nonterminal like this.

Actually, we do. Not sure about "typically", but I'd say the two approaches are roughly on par. For example, see the [definition of HasName for the same production](https://tc39.es/ecma262/#sec-class-definitions-static-semantics-hasname).

Reply:
Ah, thanks for pointing that out. I guess it doesn't much matter, then; while my preference is to avoid these splits, it could always be cleaned up later.

Reply:
I would love to get these normalised. My preference would actually be the opposite of @bakkot, but that would probably lead to a lot of duplication of algorithm steps, so it's probably best to go with @bakkot's suggestion

Reply:
Sounds like this should land as-is for now, and we can discuss normalization in a followup.